### PR TITLE
reindex/renumbering mesh layer

### PR DESCRIPTION
**Pull request by vcloarec**
This PR allows the user, during editing, to reindex (or renumbering) the vertices and the faces.
The renumbering is an optimization of the mesh with the Cuthill-McKee algorithm

Peek 07-09-2021 09-01

.

### CODE REVIEW

**Comment by nyalldawson** [comment]
I'm not sure of the exact use case for this, but would it be possible to make this a processing algorithm instead? It would add a lot of functionality "for free"! (Graphical models, batch execution, etc)

**Comment by vcloarec** [comment]
of course, it could be also a processing algorithm. I think it could be straightforward to use the new class with processing, it could be another feature.
But why instead ?

Indeed,  this operation makes:

1- A simple reindexation of the mesh indexes of faces and vertices. For mesh editing and to allow undo/redo, QGIS keeps empty place of elements that are not here any more. Many operation creates these "holes" in the container and the count of them increase rapidly. The reindexation consist to remove theses holes and reindex the elements. Even if this operation is done before each saving on file, this could be interesting to do it during edition without save, mainly to save memory. it could be seen as a purge.
2- clean the undo/redo stack, could be interesting for same reasons.
These two things are specific to QGIS, but the not the last and the more important here:
3- the renumbering with the Cuthill-McKee algorithm, this is a reindexation of the elements top optimize relation between faces and vertices. Sayid quickly, it is to avoid a face with index '4' on the side of the mesh and a face with an index '5' on the other side. This is use for finite element calculation on unstructured mesh to increase the efficiency of calculation. **_Typically this is done after an edition_** of the mesh that could break the optimization of face/vertices relation, so before saving the mesh after edition.

So this function is quite link to mesh edition that are done here only on a mesh layer on edit mode, that seems to me a natural position next other edition tools.

Replace this implementation by a processing, IMHO, will lead to a less accessible function. But I am 100% agree that the Cuthill-McKee algorithm could also be on the processing part, especially with these new  classes that could be easily reused.
But that will ask another implementation that could be done later (by me or another). Only the part 3 (Cuthill-McKee algorithm) will be applied on mesh layer not in edit mode.

**Review by PeterPetrik — COMMENTED** [review]
I think that the main usecase is when you want to persist your (edited) mesh to the file and you want to renumber the indices of faces/vertices so they are continuous and somewhat reasonable ordered (==renumbering).
All others is in my opinion too much detailed for user, so I would not announce it anyhow (like redo/undo stack, recreating of the internal mesh index, etc...)

I am not sure if we need a button for it, probably processing alg is enough? Or user should be asked on save of edited mesh if to renumber or not?

**Comment by PeterPetrik** [comment]
Let me merge this to go to 3.22. We will try to add this also as a processing algorithm and think about improving UX in the following period after FF.

**Review by PeterPetrik — APPROVED** [review]

File: src/app/mesh/qgsmaptooleditmeshframe.cpp [hunk 1]
```diff
@@ -1358,6 +1364,31 @@ void QgsMapToolEditMeshFrame::triggerTransformCoordinatesDockWidget( bool checke
...
+    return;
+
+  onEditingStarted();
+
+  if ( QMessageBox::question( canvas(), tr( "Reindex the Mesh" ),
+                              tr( "This action will reindex the faces and vertices of the mesh layer %1."
+                                  "This opration will consist to:\n\n"
+                                  "- purge memory of remaining traces of old vertices and faces\n"
+                                  "- empty the undo/redo stack of this layer\n"
+                                  "- optimize the indexes of vertices and faces using the CutHill-McKee Algorithm.\n\n"
```
[the remark this answers is not in the document]
Comment (DelazJ) [reply src/app/mesh/qgsmaptooleditmeshframe.cpp]: > or just ask if really want to reindex mesh?

Late on this but i'm wondering: if this is not expensive and does not have any chances (?) to break anything in the mesh, do we need another step of confirmation? What about just proceed and display a message of success. I see it like someone hitting F5

[the remark this answers is not in the document]
Comment (PeterPetrik) [reply src/app/mesh/qgsmaptooleditmeshframe.cpp]: I let Vincent comment of speed, but in general I would say it is not fast. For bigger meshes (like milions of cells) just creation of the index takes 1/2 of the loading time, even few seconds. Not sure how fast is the "reindexing" algorithm, but I think it is safe to assume is can take some time.

**Pull request merged**